### PR TITLE
implement path join in rust

### PR DESCRIPTION
**Pull request by ashleygwilliams**

### CODE REVIEW

**Comment by creationix** [comment]
The basic goal is something like:

``` c

const char* join(const char* a...);

const char* dirname(const char* path); -> basename

const char* extension(const char* path);

const char* filename(const char* path);
```

**Comment by steveklabnik** [comment]
TODOs:
- get rid of unwrap using http://doc.rust-lang.org/stable/std/panic/fn.catch_unwind.html . right now, if an error happens, those unwraps invoke UB.
- figure out the issue with allocation. Right now, this will leak the `CString`s, and you can't call `free()` on them, since they use a different allocator. @creationix , you didn't indicate ownership semantics, just the type signatures, I'm assuming you're assuming that these return new strings that it's the callers' job to deal with?
- figure out varargs. Rust can call C vararg functions, but I'm not sure that we have the ability to expose vararg'd functions My docs are bad. Should investigate this. I have some ideas of mitigation strategies...
- Right now, there's a lot of overhead from going to one type to another. This will happen with little C functions, but beyond that, it's possible that we could get away with different coercions here. Need to investigate these APIs a bit better, I'm not mega mega familliar/expert with this part of the stdlib yet
- This will only really work for filenames that can be converted to UTF-8. Dunno how much that stuff matters, but that might need dealing with.

I think that's all I have for now.

**Comment by creationix** [comment]
@steveklabnik I'm not particular to who owns the string data.  I'm going to copy it somewhere as soon as I get it anway.

In other words, feel free to suggest another interface that is easier/safer for rust.

File: implementations/duktape/main.c [hunk 1]
```diff
@@ -579,20 +472,21 @@ int main(int argc, char *argv[]) {
     }
   }
 
+
```
Comment (steveklabnik) [inline implementations/duktape/main.c]: extra whitespace

File: implementations/duktape/rust-path.h [hunk 2]
```diff
@@ -0,0 +1,6 @@
+
+
```
Comment (steveklabnik) [inline implementations/duktape/rust-path.h]: extra whitespace

File: implementations/duktape/rust-path.h [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+const char* rust_join(const char* a, const char* b);
+const char* rust_dirname(const char* path);
+const char* rust_extension(const char* path);
+const char* rust_filename(const char* path);
```
Comment (steveklabnik) [inline implementations/duktape/rust-path.h:4]: i'll just leave this here https://crates.io/crates/rusty-cheddar :wink:

**Comment by steveklabnik** [comment]
It's like, way too late here, so I'm going to go to sleep. But I just wanted to drop this as well: https://github.com/carols10cents/rust-out-your-c-talk/blob/master/slides-with-speaker-notes.pdf @carols10cents basically just gave a talk about doing exactly this

**Comment by steveklabnik** [comment]
To reemphasize my TODOs:

> Right now, this will leak the CStrings, and you can't call free() on them, since they use a different allocator. 

So, merging this means that you're now leaking memory. :wink:

**Comment by steveklabnik** [comment]
also going to bed for reals, i'll see what you all come up with and comment some more on the other stuff tomorrow.

**Comment by creationix** [comment]
@steveklabnik yes, I know it's leaking.  This is fine, my plan is to sometime soon replace all the C code with rust at which point we can handle the memory correctly.  (I mean duktape and libuv will still contain C obviously, but this repo will be 100% rust using a couple crates that happen to contain C)

See also https://twitter.com/creationix/status/740367352533909504

**Comment by creationix** [comment]
@steveklabnik yes, I know it's leaking.  This is fine, my plan is to sometime soon replace all the C code with rust at which point we can handle the memory correctly.  (I mean duktape and libuv will still contain C obviously, but this repo will be 100% rust using a couple crates that happen to contain C)

See also https://twitter.com/creationix/status/740367352533909504

**Comment by steveklabnik** [comment]
👍👍👍

Btw, PR incoming for some other stuff.

**Comment by carols10cents** [comment]
> It's like, way too late here, so I'm going to go to sleep. But I just wanted to drop this as well: https://github.com/carols10cents/rust-out-your-c-talk/blob/master/slides-with-speaker-notes.pdf @carols10cents basically just gave a talk about doing exactly this

HI HI HI HI great minds think alike!!!!! I definitely left some details out of the talk, and I know I haven't understood all the nuances everywhere yet, but I'd love to help what yinz are doing here if I can be of use!!

**Comment by creationix** [comment]
Oh hi @carols10cents.  I'm traveling today and will be at an on-site with my work team all week.  But I might sneak in some time to work on this if I can.  I've downloaded the pdf to go through later.

**Pull request merged**